{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0059eb9e-e490-4dbf-bb7c-baa2539612b9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4580d2be-1621-4098-8cfd-934f97cc456d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f03294b-f265-47c1-a46f-03630e1768a7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "696ade42-7add-4d95-8c2f-9eec2954a844",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2dcf22a4-327f-4b80-a20a-a352cba38be0",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "657fa2ba-db1b-4eaf-a093-b40d64686fa2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SpaceRanger:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1187ed45-0b14-4c33-8119-2e697a35dd30",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ./run_spaceranger.single.sh p96_p1_run3 P96_P1 P96_P1 Visium.image/jpg/P96P_updated.jpeg V12D14-351 D1 /genomics/1_Projects/Ecig_U01/Visium/visium_final_count/96P_json/V12D14-351-D1.json\n",
    "\n",
    "# using \n",
    "\n",
    "#./spaceranger-2.1.1/spaceranger count --id  <Folder path to store results> --transcriptome <Folder path containing Rat transcriptome information and reference in JSON format> --fastqs <Folder path containing fastq files> --sample <sample filename prefix> --include-introns true --localmem 300 --localcores 60 --image <File path of the slice/sample figure in JPG or TIFF format> --slide <slide code name> --area <area code> --loupe-alignment <File path of the Loupe alignment in JSON format>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1812b8c4-2b21-41dd-975d-fe97ef1de61f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ca3ba39-0f02-44b5-8499-f28a2fa766a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Velocyto:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c48c8256-0c14-472f-b467-31a16b8ad8e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# velocyto run10x -@ 60 --samtools-memory 100000 <Folder path to SpaceRanger results> <File path of the Rat genes gtf>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f928684-813a-4e4f-9e12-8f44938e6ea9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data cleaning and Integration "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5df7f973-a7d7-4819-be84-0a6b84a5157f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scvi\n",
    "import os\n",
    "from scipy.sparse import csr_matrix\n",
    "import diffxpy.api as de\n",
    "\n",
    "PATH_TO_INPUT_DATA=\"/genomics/Rachid/Velocity/\"\n",
    "FILE_VELO_ADATA=\"velo_mp_integrated.July16.h5ad\"\n",
    "MODEL_VELO_ADATA=\"model_integrated_velo_mp.July16.model\"\n",
    "\n",
    "d_sample_to_dirname = {\n",
    "        \"P94P1\":\"p94_p1_run2\",\n",
    "        \"P95P1\":\"p95_p1_run2\",\n",
    "        \"P100P1\":\"p100_p1_run2\",\n",
    "        \"P81P2\":\"p81_p2_run2\",\n",
    "        \"P81P1\":\"p81_p1_run2\",\n",
    "        \"P83P1\":\"p83_p1_run2\",\n",
    "        \"P112P1\":\"p112_p1_run2\",\n",
    "        \"P113P1\":\"p113_p1_run2\",\n",
    "        \"P80P1\":\"p80_p1_run2\",\n",
    "        }\n",
    "\n",
    "d_sample_to_categ = {\n",
    "        \"P94P1\":\"CM\",\n",
    "        \"P95P1\":\"CM\",\n",
    "        \"P100P1\":\"CM\",\n",
    "        \"P81P2\":\"EcigM\",\n",
    "        \"P81P1\":\"EcigM\",\n",
    "        \"P83P1\":\"EcigM\",\n",
    "        \"P112P1\":\"EcigM\",\n",
    "        \"P113P1\":\"EcigM\",\n",
    "        \"P80P1\":\"EcigM\",\n",
    "        }\n",
    "\n",
    "def preprocess_data(folder_name, loom_file, idx_sample):\n",
    "    adata = sc.read_10x_mtx(folder_name)\n",
    "    sc.pp.filter_cells(adata, min_genes=200)\n",
    "    adata.var['mt'] = adata.var_names.str.startswith('Mt-')\n",
    "    sc.pp.calculate_qc_metrics(adata, qc_vars=[\"mt\"], percent_top=None, log1p=False, inplace=True)\n",
    "    upper_lim = np.quantile(adata.obs.n_genes_by_counts.values, .98)\n",
    "    lower_lim = np.quantile(adata.obs.n_genes_by_counts.values, .02)\n",
    "    upper_lim_pct = np.quantile(adata.obs.pct_counts_mt.values, .98)\n",
    "    lower_lim_pct = np.quantile(adata.obs.pct_counts_mt.values, .02)\n",
    "    print(f\"genes_by_counts: {lower_lim} to {upper_lim}, pct_mt: {lower_lim_pct} to {upper_lim_pct}\")\n",
    "    print(folder_name + \": Original size: \" + str(len(adata)))\n",
    "    adata = adata[(adata.obs.n_genes_by_counts < upper_lim) & (adata.obs.n_genes_by_counts > lower_lim)]\n",
    "    print(folder_name + \": After filtering per n_genes_by_counts: \" + str(len(adata)))\n",
    "    adata = adata[adata.obs.pct_counts_mt < upper_lim_pct, :]\n",
    "    print(folder_name + \": After filtering per pct_counts_mt: \" + str(len(adata)))\n",
    "\n",
    "    ldata = scv.read(loom_file)\n",
    "    adata = scv.utils.merge(adata, ldata)\n",
    "    scv.pp.filter_and_normalize(adata)\n",
    "    scv.pp.moments(adata)\n",
    "    scv.tl.velocity(adata)\n",
    "    adata.obs.pop(\"initial_size_unspliced\")\n",
    "    adata.obs.pop(\"initial_size_spliced\")\n",
    "    adata.obs.pop(\"initial_size\")\n",
    "    adata.obs[\"Sample\"] = idx_sample\n",
    "    return adata\n",
    "\n",
    "def run_integration():\n",
    "    if not os.path.exists(FILE_VELO_ADATA):\n",
    "        adata_AF = []\n",
    "        for (s_name, s_dirname) in d_sample_to_dirname.items():\n",
    "            print(\"Adding \" + s_name)\n",
    "            adata_AF.append(\n",
    "                    preprocess_data(\n",
    "                        PATH_TO_INPUT_DATA + s_dirname + \"/outs/filtered_feature_bc_matrix/\",\n",
    "                        PATH_TO_INPUT_DATA + s_dirname + \"/velocyto/\" + s_dirname +\".loom\" ,\n",
    "                        s_name\n",
    "                        )\n",
    "                    )\n",
    "        adata = sc.concat(adata_AF)\n",
    "        adata.obs[\"Condition\"] = adata.obs.Sample.map(d_sample_to_categ)\n",
    "\n",
    "        sc.pp.filter_genes(adata, min_cells=200)\n",
    "\n",
    "        layers = [\"spliced\", \"unspliced\", \"ambigious\"]\n",
    "        layers_keys = [key for key in layers if key in adata.layers.keys()]\n",
    "        counts_layers = [np.sum(adata.layers[key], axis=1) for key in layers_keys]\n",
    "        counts_total = np.sum(counts_layers, 0)\n",
    "        counts_total += counts_total == 0\n",
    "        counts_layers = np.array([counts / counts_total for counts in counts_layers])\n",
    "        adata.obs[\"U_US_ratio\"] = counts_layers[1]\n",
    "\n",
    "        adata.X = csr_matrix(adata.X)\n",
    "        adata.obs.groupby(\"Sample\").count()\n",
    "        # Saving Raw data\n",
    "        adata.layers[\"counts\"] = adata.X.copy()\n",
    "        adata.raw_unnormalized = adata.copy()\n",
    "\n",
    "        sc.pp.normalize_total(adata)\n",
    "        sc.pp.log1p(adata)\n",
    "        sc.pp.highly_variable_genes(adata)\n",
    "\n",
    "        adata.raw = adata.copy()\n",
    "\n",
    "        print(\"Running Integration of samples\")\n",
    "        # Integration method:\n",
    "        scvi.model.SCVI.setup_anndata(adata, layer = \"counts\",\n",
    "                categorical_covariate_keys = [\"Sample\"],\n",
    "                continuous_covariate_keys = [\"pct_counts_mt\", \"n_counts\"])\n",
    "        model = scvi.model.SCVI(adata)\n",
    "        model.train()\n",
    "        model.get_latent_representation().shape\n",
    "        adata.obsm[\"X_scVI\"] = model.get_latent_representation()\n",
    "        sc.pp.neighbors(adata, use_rep = \"X_scVI\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e5b1949-ef0f-420e-849f-62a360dbc906",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merging RNA velocity data from each sample into one integrated AnnData instance:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c40711ec-1806-477e-a4dd-fbe3838e68ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import scvelo as scv\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "scv.settings.verbosity = 3  # show errors(0), warnings(1), info(2), hints(3)\n",
    "scv.settings.presenter_view = True  # set max width size for presenter view\n",
    "scv.set_figure_params('scvelo')  # for beautified visualization\n",
    "\n",
    "MALEDATATYPE=\"fp\"\n",
    "FILE_VELO_MA_EXT_ADATA=\"./velo_\" + MALEDATATYPE + \"_integrated.July16.h5ad\"\n",
    "FILE_VELO_UPD_M_ADATA_PRE=\"./VersionJuly1/velo_\"+ MALEDATATYPE + \"_anndata/\"\n",
    "\n",
    "\n",
    "def merge_anndata_objects(intg_adata_ext, ldata):\n",
    "    try:\n",
    "        adata_vz = scv.utils.merge(intg_adata_ext, ldata)\n",
    "    except AttributeError as e:\n",
    "        # Custom debug message for AttributeError\n",
    "        print(f\"AttributeError encountered: {e}\")\n",
    "        # Detailed debugging\n",
    "        for k in intg_adata_ext.obs:\n",
    "            print(f\"Column: {k}, Dtype: {intg_adata_ext.obs[k].dtype}\")\n",
    "\n",
    "        for k in ldata.obs:\n",
    "            print(f\"Column: {k}, Dtype: {ldata.obs[k].dtype}\")\n",
    "        raise\n",
    "    return adata_vz\n",
    "\n",
    "\n",
    "def rotate_coordinates(x, y, angle_degrees = 180):\n",
    "    # Ensure x and y are numpy arrays\n",
    "    x = np.array(x)\n",
    "    y = np.array(y)\n",
    "\n",
    "    # Calculate the midpoint of the dataset on the x-axis and y-axis\n",
    "    midpoint_x = np.mean(x)\n",
    "    midpoint_y = np.mean(y)\n",
    "\n",
    "    # Convert the angle from degrees to radians\n",
    "    angle_radians = np.radians(angle_degrees)\n",
    "\n",
    "    # Create the rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [np.cos(angle_radians), -np.sin(angle_radians)],\n",
    "        [np.sin(angle_radians), np.cos(angle_radians)]\n",
    "        ])\n",
    "\n",
    "    # Center the points around the midpoint\n",
    "    centered_points = np.vstack((x - midpoint_x, y - midpoint_y))\n",
    "\n",
    "    # Rotate the points\n",
    "    rotated_points = rotation_matrix.dot(centered_points)\n",
    "\n",
    "    # Shift the points back to the original midpoint\n",
    "    rotated_x = rotated_points[0, :] + midpoint_x\n",
    "    rotated_y = rotated_points[1, :] + midpoint_y\n",
    "\n",
    "    # Convert the rotated coordinates back to lists\n",
    "    return rotated_x.tolist(), rotated_y.tolist()\n",
    "\n",
    "\n",
    "def generate_ann_data_sample(folder_name, sample_code, fig_sample_name, coord_filename):\n",
    "    print(\"START: Generating AnnData for \" + fig_sample_name + \" ... \")\n",
    "    intg_adata = sc.read_h5ad(FILE_VELO_MA_EXT_ADATA)\n",
    "    path_prefix_file = \"./\" + folder_name + \"/\" + sample_code + \"_run2\"\n",
    "\n",
    "# Read the CSV file containing morphological annotations for each spot:\n",
    "    csv_file = pd.read_csv(\"morph_mapping/\" + coord_filename)\n",
    "\n",
    "# Extract the first column (index column)\n",
    "    index_column = csv_file.iloc[:, 0].astype(str).str[:12]\n",
    "    intg_adata_proj = intg_adata[intg_adata.obs.Sample == fig_sample_name]\n",
    "    filtered_adata = intg_adata_proj[intg_adata_proj.obs_names.isin(index_column)]\n",
    "    filtered_adata_df = filtered_adata.to_df()\n",
    "    csv_file = csv_file.set_index(csv_file.columns[0])\n",
    "\n",
    "# Use the first 12 characters of the CSV index to match the filtered AnnData index\n",
    "    csv_file.index = csv_file.index.astype(str).str[:12]\n",
    "    result_df = csv_file.join(filtered_adata_df, how='inner')\n",
    "    result_df_col6 = result_df[[\"row\", \"col\", \"imagerow\", \"imagecol\", \"region\"]]\n",
    "\n",
    "# Ensure both have the same index (common key)\n",
    "    intg_adata_proj.obs_names = intg_adata_proj.obs_names.astype(str)\n",
    "    result_df_col6.index = result_df_col6.index.astype(str)\n",
    "\n",
    "# Subset the AnnData object to include only rows that match the metadata DataFrame's index\n",
    "    intg_adata_ext = intg_adata_proj[intg_adata_proj.obs_names.isin(result_df_col6.index)]\n",
    "\n",
    "# Add the metadata columns to the AnnData object's obs\n",
    "    for col in result_df_col6.columns:\n",
    "        intg_adata_ext.obs[col] = result_df_col6.loc[intg_adata_ext.obs_names, col]\n",
    "\n",
    "    intg_adata_ext.obs.rename(columns={\"region\": \"Regions\"}, inplace=True)\n",
    "    ldata = scv.read(path_prefix_file + \".loom\")\n",
    "\n",
    "    adata_vz =  merge_anndata_objects(intg_adata_ext, ldata)\n",
    "\n",
    "    positions = pd.read_csv(\"./SpaceRangerTissuePositions/\" + sample_code + \"_run2tissue_positions.csv\")\n",
    "    positions = positions.tail(positions.shape[0] -1)\n",
    "    positions.index = positions.index.rename(\"CellID\")\n",
    "    positions[\"CellID\"] = [a[:-6] for a in positions[\"barcode\"]]\n",
    "    positions.drop(\"barcode\", inplace=True, axis=1)\n",
    "    positions.index = positions[\"CellID\"]\n",
    "    positions['pxl_col_in_fullres'] = positions['pxl_col_in_fullres'].astype('float32')\n",
    "    positions['pxl_row_in_fullres'] = positions['pxl_row_in_fullres'].astype('float32')\n",
    "\n",
    "    new_x = positions['pxl_row_in_fullres']\n",
    "    new_y = positions['pxl_col_in_fullres']\n",
    "\n",
    "    if fig_sample_name in [\"P93A1\", \"P81A2\", \"P100P1\", \"P112P1\", \"P113P1\", \"P112A1\", \"P100A1\"]:\n",
    "        print(\"transforming grid coords...\")\n",
    "        new_x_2, new_y_2 = rotate_coordinates(new_x, new_y)\n",
    "\n",
    "    positions['pxl_row_in_fullres'] = new_x_2 \n",
    "    positions['pxl_col_in_fullres'] = new_y_2\n",
    "\n",
    "    adata_vz.obs = adata_vz.obs.join(positions, how=\"left\")\n",
    "\n",
    "    adata_vz.obsm['X_umap'] = adata_vz.obs[\n",
    "            ['pxl_row_in_fullres', 'pxl_col_in_fullres']\n",
    "            ].to_numpy()\n",
    "\n",
    "    warnings.filterwarnings('ignore')\n",
    "    scv.pp.filter_and_normalize(adata_vz)\n",
    "    scv.pp.moments(adata_vz)\n",
    "\n",
    "    warnings.filterwarnings('ignore')\n",
    "    sc.pp.neighbors(adata_vz)\n",
    "    scv.tl.velocity(adata_vz)\n",
    "    scv.tl.velocity_graph(adata_vz)\n",
    "    scv.set_figure_params()\n",
    "    scv.tl.recover_dynamics(adata_vz, n_jobs = 40)\n",
    "    scv.tl.velocity(adata_vz, mode=\"dynamical\")\n",
    "    scv.tl.velocity_graph(adata_vz)\n",
    "    scv.pl.velocity_embedding_stream(adata_vz, basis = \"umap\", color = \"Regions\", save= \"_intg_\" + fig_sample_name +\"_dyn_stream\")\n",
    "    scv.tl.latent_time(adata_vz)\n",
    "    scv.pl.scatter(adata_vz, color = \"latent_time\", color_map = \"gnuplot\", size=80, save=  \"_intg_\" + fig_sample_name + \"_latent_time\")\n",
    "    warnings.filterwarnings('ignore')\n",
    "    scv.pl.velocity_embedding_grid(adata_vz, basis = \"umap\", color = \"Regions\", arrow_length=1, arrow_size=1, dpi=120, save = \"_intg_\" + fig_sample_name + \"_grid_velo\")\n",
    "\n",
    "    warnings.filterwarnings('ignore')\n",
    "    scv.tl.velocity_pseudotime(adata_vz)\n",
    "    scv.pl.scatter(adata_vz, color='velocity_pseudotime', cmap='gnuplot', save = \"_intg_\" + fig_sample_name + \"_velo_pseudo_time\")\n",
    "\n",
    "    scv.tl.velocity_confidence(adata_vz)\n",
    "    keys = \"velocity_length\", \"velocity_confidence\"\n",
    "    scv.pl.scatter(adata_vz, c = keys, perc = [5, 95],  save = \"_intg_\" + fig_sample_name + \"_velo_length_confidence\")\n",
    "    scv.pl.scatter(adata_vz, color=\"U_US_ratio\", save = \"_intg_\" + fig_sample_name + \"_u_us_ratio\")\n",
    "\n",
    "    adata_vz.write_h5ad(FILE_VELO_UPD_M_ADATA_PRE + fig_sample_name + \".h5ad\")\n",
    "\n",
    "    print(\"END: AnnData for \" + fig_sample_name + \" created.\")\n",
    "\n",
    "if False:\n",
    "    generate_ann_data_sample(\"fa_velocyto_loom\", \"p101_a1\", \"P101A1\", \"FA101.cor.csv\")\n",
    "    generate_ann_data_sample(\"fa_velocyto_loom\", \"p108_a1\", \"P108A1\", \"FA108.cor.csv\")\n",
    "    generate_ann_data_sample(\"fa_velocyto_loom\", \"p111_a1\", \"P111A1\", \"FA111.cor.csv\")\n",
    "    generate_ann_data_sample(\"fa_velocyto_loom\", \"p74_a1\", \"P74A1\", \"FA74.cor.csv\")\n",
    "    generate_ann_data_sample(\"fa_velocyto_loom\", \"p77_a1\", \"P77A1\", \"FA77.cor.csv\")\n",
    "    generate_ann_data_sample(\"fa_velocyto_loom\", \"p82_a1\", \"P82A1\", \"FA82.cor.csv\")\n",
    "    generate_ann_data_sample(\"fa_velocyto_loom\", \"p93_a1\", \"P93A1\", \"FA93.cor.csv\")\n",
    "    generate_ann_data_sample(\"fa_velocyto_loom\", \"p96_a1\", \"P96A1\", \"FA96.cor.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "392c99ae-d7de-4d26-a201-2360377b42b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Differential Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8c366b31-fae6-4dc8-8af1-6a300ce90578",
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import math\n",
    "import scipy\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from collections import defaultdict\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import scvelo as scv\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import scvi\n",
    "import os\n",
    "from scipy.sparse import csr_matrix\n",
    "import diffxpy.api as de\n",
    "\n",
    "#FA\n",
    "FILE_VELO_FA_COMP_ADATA=\"velo_fa_integrated.completed.July10.genes_velo.h5ad\"\n",
    "FILE_VELO_FA_ADATA=\"velo_fa_integrated.completed.July1.genes.h5ad\"\n",
    "MODEL_VELO_FA_COMP_ADATA=\"model_integrated_velo_fa.completed.July10.genes_velo.model\"\n",
    "#MA\n",
    "FILE_VELO_MA_COMP_ADATA=\"velo_ma_integrated.completed.July11.genes.h5ad\"\n",
    "FILE_VELO_MA_ADATA=\"velo_ma_integrated.completed.July11.h5ad\"\n",
    "MODEL_VELO_MA_COMP_ADATA=\"model_integrated_velo_ma.completed.July11.genes.model\"\n",
    "\n",
    "#MP\n",
    "FILE_VELO_MP_COMP_ADATA=\"velo_mp_integrated.completed.July16.genes_velo.h5ad\"\n",
    "FILE_VELO_MP_ADATA=\"velo_mp_integrated.completed.July16.h5ad\"\n",
    "MODEL_VELO_MP_COMP_ADATA=\"model_integrated_velo_mp.completed.July16.genes_velo.model\"\n",
    "\n",
    "#FP\n",
    "FILE_VELO_FP_COMP_ADATA=\"velo_fp_integrated.completed.July16.genes_velo.h5ad\"\n",
    "FILE_VELO_FP_ADATA=\"velo_fp_integrated.completed.July16.h5ad\"\n",
    "MODEL_VELO_FP_COMP_ADATA=\"model_integrated_velo_fp.completed.July16.genes_velo.model\"\n",
    "\n",
    "def run_integration():\n",
    "    if not os.path.exists(FILE_VELO_MP_COMP_ADATA):\n",
    "        adata = sc.read_h5ad(FILE_VELO_MP_ADATA)\n",
    "        # Removal of cells with no recognized regions\n",
    "        vals = adata.obs['Regions'].unique()\n",
    "        res = []\n",
    "        for v in range(len(vals)):\n",
    "            if len(vals[v]) > 8:\n",
    "                res.append(vals[v])\n",
    "        adata = adata[adata.obs['Regions'].isin(res) == False]\n",
    "        adata.obs.drop(\n",
    "            columns=[\"leiden\", \"_scvi_batch\", \"_scvi_labels\", \"pxl_col_in_fullres\", \"pxl_row_in_fullres\"],\n",
    "            inplace=True,\n",
    "        )\n",
    "        # Remove Velocity-based attribbutes\n",
    "        #adata.obs.drop(\n",
    "        #    columns=[\"latent_time\", \"velocity_length\", \"velocity_pseudotime\", \"U_US_ratio\"],\n",
    "        #    inplace=True,\n",
    "        #)\n",
    "        adata.X = csr_matrix(adata.X)\n",
    "        adata.obs.groupby(\"Sample\").count()\n",
    "        # Saving Raw data\n",
    "        adata.layers[\"counts\"] = adata.X.copy()\n",
    "        adata.raw_unnormalized = adata.copy()\n",
    "\n",
    "        #sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "        sc.pp.normalize_total(adata)\n",
    "        sc.pp.log1p(adata)\n",
    "        sc.pp.highly_variable_genes(adata)\n",
    "        print(\"We keep highly variable genes, size:\")\n",
    "        print(len(adata.obs))\n",
    "        print(len(adata.var))\n",
    "\n",
    "        adata.raw = adata.copy()\n",
    "\n",
    "        print(\"Running Integration of samples\")\n",
    "        # Integration method:\n",
    "        scvi.model.SCVI.setup_anndata(\n",
    "                adata,\n",
    "                layer = \"counts\",\n",
    "                categorical_covariate_keys = [\"Sample\"],\n",
    "                continuous_covariate_keys = [\"pct_counts_mt\", \"n_counts\", \"latent_time\", \"velocity_length\", \"velocity_pseudotime\", \"U_US_ratio\"]\n",
    "                )\n",
    "        model = scvi.model.SCVI(adata)\n",
    "        model.train()\n",
    "        model.get_latent_representation().shape\n",
    "        adata.obsm[\"X_scVI\"] = model.get_latent_representation()\n",
    "\n",
    "        adata.write_h5ad(FILE_VELO_MP_COMP_ADATA)\n",
    "        model.save(MODEL_VELO_MP_COMP_ADATA)\n",
    "        return adata, model\n",
    "\n",
    "# adata, model run_integration()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e068efba-e218-48f3-bc4d-6d6b2e558348",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import scvelo as scv\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import math\n",
    "import scipy\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy import stats\n",
    "\n",
    "from collections import defaultdict\n",
    "import warnings\n",
    "from scipy.stats import chisquare\n",
    "warnings.filterwarnings('ignore')\n",
    "scv.settings.verbosity = 3  # show errors(0), warnings(1), info(2), hints(3)\n",
    "scv.settings.presenter_view = True  # set max width size for presenter view\n",
    "scv.set_figure_params('scvelo')  # for beautified visualization\n",
    "\n",
    "import diffxpy.api as de\n",
    "import scvi\n",
    "import os\n",
    "from scipy.sparse import csr_matrix\n",
    "from scipy import stats\n",
    "\n",
    "FILE_VELO_FP_CMP_GV_ADATA = \"../FP/velo_fp_integrated.completed.July16.genes_velo.h5ad\"\n",
    "intg_adata_gv = sc.read_h5ad(FILE_VELO_FP_CMP_GV_ADATA)\n",
    "MODEL_VELO_FP_CMP_GV_ADATA = \"../FP/model_integrated_velo_fp.completed.July16.genes_velo.model/\"\n",
    "model = scvi.model.SCVI.load(MODEL_VELO_FP_CMP_GV_ADATA, intg_adata_gv)\n",
    "\n",
    "sc.pp.normalize_total(intg_adata_gv)\n",
    "sc.pp.log1p(intg_adata_gv)\n",
    "sc.pp.highly_variable_genes(intg_adata_gv, n_top_genes=2000)\n",
    "sc.tl.rank_genes_groups(intg_adata_gv, \"Regions\", method='wilcoxon')\n",
    "sc.pl.rank_genes_groups(intg_adata_gv, n_genes=25, sharey=False)\n",
    "\n",
    "def find_differentially_veloce_genes(l_adata, l_markers_scvi, name, control_code, treatment_code, min_pvalue = 0.05):\n",
    "    lcl_subset = l_adata[l_adata.obs[\"Regions\"].isin([name])].copy()\n",
    "    lcl_subset.X = lcl_subset.X.toarray()\n",
    "    markers_scvi_lcl = l_markers_scvi[l_markers_scvi[\"group1\"] == name]\n",
    "    tmp = lcl_subset[lcl_subset.obs.Regions == name]\n",
    "    for g_name in markers_scvi_lcl.index:\n",
    "        i = np.where(tmp.var_names == g_name)[0][0]\n",
    "        a = tmp[tmp.obs.Condition == treatment_code].X[:,i]\n",
    "        b = tmp[tmp.obs.Condition == control_code].X[:,i]\n",
    "        pval = stats.mannwhitneyu(a, b).pvalue\n",
    "        if pval < min_pvalue:\n",
    "            print(\"region,\" + name + \",p-value,\" + str(pval) + \",\" + g_name + \",proba_de,\" + str(markers_scvi_lcl.at[g_name, \"proba_de\"]))\n",
    "\n",
    "def run_differential_gene_kinetics(lcl_adata, lcl_model, control_code, treatment_code):  \n",
    "    markers_1 = sc.get.rank_genes_groups_df(lcl_adata, None)\n",
    "    markers_1 = markers_1[(markers_1.pvals_adj < 0.05) & (markers_1.logfoldchanges > 0.25)]\n",
    "    markers_scvi_1 = lcl_model.differential_expression(groupby = \"Regions\")\n",
    "    markers_scvi_1 = markers_scvi_1[(markers_scvi_1[\"is_de_fdr_0.05\"]) & (markers_scvi_1.lfc_mean > 0.5)]\n",
    "\n",
    "#    for name in [\"cc\", \"CPu\", \"NAc\", \"AIC\", \"Septal\", \"Pir\", \"Sparse\", \"ACC\", \"LV\", \"Layer1\", \"Layer2/3\", \"Layer4\", \"Layer5/6\"]:    \n",
    "    for name in [\"HY\",\"Layer2/3\",\"cc\",\"Amy\",\"Layer4\",\"Layer5/6\",\"TH\",\"Sparse\",\"HP\",\"LV\",\"Layer1\",\"CPu\",\"RSP\"]:\n",
    "        find_differentially_veloce_genes(lcl_adata, markers_scvi_1, name, control_code, treatment_code, 0.05)\n",
    "\n",
    "# For female data:\n",
    "run_differential_gene_kinetics(intg_adata_gv ,model, \"CF\", \"EcigF\")\n",
    "# run_differential_gene_kinetics(intg_adata_gv ,model, \"CM\", \"EcigM\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
